runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/control/FnApiControlClient.java

```diff
@@ -53,9 +55,9 @@
   private final ResponseStreamObserver responseObserver = new ResponseStreamObserver();
   private final ConcurrentMap<String, CompletableFuture<BeamFnApi.InstructionResponse>>
       outstandingRequests;
+  private final Collection<Consumer<FnApiControlClient>> onCloseListeners = new ArrayList<>();
```

Comment:
This needs to be a thread-safe collection, as the rest of the class is written to handle concurrent access.

Reply:
Made the changes.